w3af/plugins/crawl/url_payment_finder.py

```diff
@@ -0,0 +1,219 @@
+"""
+url_payment_finder.py
+
+Copyright 2006 Andres Riancho
+
+This file is part of w3af, http://w3af.org/ .
+
+w3af is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation version 2 of the License.
+
+w3af is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with w3af; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+"""
+from itertools import chain, repeat, izip
+
+import w3af.core.controllers.output_manager as om
+import w3af.core.data.kb.knowledge_base as kb
+from w3af.core.controllers.plugins.crawl_plugin import CrawlPlugin
+from w3af.core.controllers.core_helpers.fingerprint_404 import is_404
+from w3af.core.data.dc.headers import Headers
+from w3af.core.data.bloomfilter.scalable_bloom import ScalableBloomFilter
+from w3af.core.data.parsers.doc.url import URL
+from w3af.core.data.kb.info import Info
+from w3af.core.data.request.fuzzable_request import FuzzableRequest
+
+
+class url_payment_finder(CrawlPlugin):
+    """
+    Try to find hidden gateway payment confirmation page.
+    :author: Coiffey Pierre (pierre.coiffey@gmail.com)
+    """
+    _dirs = ('/', '/inc/', '/include/', '/include/pay/', '/includes/', '/includes/pay/', '/lib/',
+             '/libraries/', '/module/', '/module/pay/', '/modules/', '/modules/pay/', '/payment/',
+             '/shop/', '/store/', '/svc/', '/servlet/', '/cgi/', '/cgi-bin/', '/cgibin/',
+    )
+
+    _files = ('pay', 'payment', 'success', 'paymentsuccess', 'paymentcomplete', 'paymentsuccessful',
+              'successful', 'paid', 'return', 'valid', 'validpay', 'validate', 'validatepayment',
+              'validatepay', 'validation', 'complete', 'completepay', 'completepayment',
+              'trxcomplete', 'transactioncomplete', 'final', 'finished',
+    )
+    _exts = ('', '.php', '.asp', '.aspx', '.jsp', '.py', 
+             '.pl', '.rb', '.cgi', '.php3',  '.php4', '.php5',
+    )
+
+    def __init__(self):
+
+        CrawlPlugin.__init__(self)
+
+        self._headers = None
+        self._first_time = True
+        self._fuzz_images = False
+        self._seen = ScalableBloomFilter()
+
+    def crawl(self, fuzzable_request):
+        """
+        Searches for new Url's using fuzzing.
+
+        :param fuzzable_request: A fuzzable_request instance that contains
+                                    (among other things) the URL to test.
+        """
+        url = fuzzable_request.get_url()
+        self._headers = Headers([('Referer', url.url_string)])
+
+        if self._first_time:
+            self._verify_head_enabled(url)
```

Comment:
The goal of sending HTTP HEAD requests in pykto was to reduce the network traffic. This plugin doesn't send that many HTTP requests, so it is fine to just send GET requests. You can go ahead and remove all the logic associated with HEAD.

The good thing about this HEAD code is that it made me think about HTTP methods! Some / most payment gateway webhooks will receive data over POST. Some web application frameworks will only route the traffic to the controller if the method (POST) is used.

So I believe that for each URL we need to try GET and POST.

Reply:
Still an issue

Reply:
I removed all the logic associated with HEAD. I did not try for the POST request though.